src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -61,20 +72,158 @@ private static void CheckIsLastArgumentCancellationTokenOrRequestContext(ISymbol
+            {
+                fullNamespace = fullNamespace == "" ? currentName : $"{currentName}.{fullNamespace}";
+                currentNamespace = currentNamespace.ContainingNamespace;
+                currentName = currentNamespace.Name;
+            }
+            return fullNamespace;
+        }
+
+        // A protocol method should not have model as parameter. If it has ambiguity with convenience method, it should have required RequestContext.
+        // Ambiguity: doesn't have a RequestContent, but there is a method ending with CancellationToken has same type of parameters 
+        // No ambiguity: has RequestContent.
+        private static void CheckProtocolMethodParameters(ISymbolAnalysisContext context, IMethodSymbol method)
+        {
+            var containsModel = method.Parameters.Any(p =>
+            {
+                var fullNamespace = GetFullNamespaceName(p);
+                return !fullNamespace.StartsWith("System") && !fullNamespace.StartsWith("Azure");
+            });
+
+            if (containsModel)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
+                return;
+            }
+
+            var requestContent = method.Parameters.FirstOrDefault(IsRequestContent);
+            if (requestContent == null && method.Parameters.Last().IsOptional)
+            {
+                INamedTypeSymbol type = (INamedTypeSymbol)context.Symbol;
+                IEnumerable<IMethodSymbol> methodList = type.GetMembers(method.Name).OfType<IMethodSymbol>().Where(member => !SymbolEqualityComparer.Default.Equals(member, method));
+                ImmutableArray<IParameterSymbol> parametersWithoutLast = method.Parameters.RemoveAt(method.Parameters.Length - 1);
+                IMethodSymbol convenienceMethod = FindMethod(methodList, method.TypeParameters, parametersWithoutLast, symbol => IsCancellationToken(symbol));
+                if (convenienceMethod != null)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0019, method.Locations.FirstOrDefault()), method);
                 }
             }
         }
 
+        // A protocol method should not have model as type. Accepted return type: Response, Task<Response>, Response<bool>, Task<Response<bool>>, Pageable<BinaryData>, AsyncPageable<BinaryData>, Operation<BinaryData>, Task<Operation<BinaryData>>, Operation, Task<Operation>, Operation<Pageable<BinaryData>>, Task<Operation<AsyncPageable<BinaryData>>>
+        private static void CheckProtocolMethodReturnType(ISymbolAnalysisContext context, IMethodSymbol method)
+        {
+            bool IsValidPageable(ITypeSymbol typeSymbol)
+            {
+                if ((!IsOrImplements(typeSymbol, PageableTypeName)) && (!IsOrImplements(typeSymbol, AsyncPageableTypeName)))
+                {
+                    return false;
+                }
```

Comment:
Emmm, I could remove it, but my initial thought is from the method `IsValidPageable` itself perspective, I should check if it is a pageable type because it is called `IsValidPageable`, otherwise if an operation type is passed mistakenly, it will still return `true`.